clients/js-legacy/src/instructions/unwrapLamports.ts

```diff
@@ -0,0 +1,156 @@
+import { TOKEN_PROGRAM_ID } from '../constants.js';
+import {
+    TokenInvalidInstructionDataError,
+    TokenInvalidInstructionKeysError,
+    TokenInvalidInstructionProgramError,
+    TokenInvalidInstructionTypeError,
+} from '../errors.js';
+import { addSigners } from './internal.js';
+import { TokenInstruction } from './types.js';
+import { COptionU64Layout } from '../serialization.js';
+
+/** TODO: docs */
+export interface UnwrapLamportsInstructionData {
+    instruction: TokenInstruction.UnwrapLamports;
+    amount: bigint | null;
+}
+
+/** TODO: docs */
+export const unwrapLamportsInstructionData = struct<UnwrapLamportsInstructionData>([
+    u8('instruction'),
+    new COptionU64Layout('amount'),
+]);
+
+/**
+ * Construct a UnwrapLamports instruction
+ *
+ * @param source       Native source account
+ * @param destination  Account receiving the lamports
+ * @param owner        Owner of the source account
+ * @param amount       Amount of lamports to unwrap
+ * @param multiSigners Signing accounts if `authority` is a multisig
+ * @param programId    SPL Token program account
+ *
+ * @return Instruction to add to a transaction
+ */
+export function createUnwrapLamportsInstruction(
+    source: PublicKey,
+    destination: PublicKey,
+    owner: PublicKey,
+    amount: bigint | null,
+    multiSigners: (Signer | PublicKey)[] = [],
+    programId: PublicKey,
+): TransactionInstruction {
+    const keys = addSigners(
+        [
+            { pubkey: source, isSigner: false, isWritable: true },
+            { pubkey: destination, isSigner: false, isWritable: true },
+        ],
+        owner,
+        multiSigners,
+    );
+
+    const data = Buffer.alloc(10); // worst-case
+    unwrapLamportsInstructionData.encode(
+        {
+            instruction: TokenInstruction.UnwrapLamports,
+            amount,
+        },
+        data,
+    );
+
+    return new TransactionInstruction({ keys, programId, data });
+}
+
+/** A decoded, valid UnwrapLamports instruction */
+export interface DecodedUnwrapLamportsInstruction {
+    programId: PublicKey;
+    keys: {
+        source: AccountMeta;
+        destination: AccountMeta;
+        owner: AccountMeta;
+        multiSigners: AccountMeta[];
+    };
+    data: {
+        instruction: TokenInstruction.UnwrapLamports;
+        amount: bigint | null;
+    };
+}
+
+/**
+ * Decode a UnwrapLamports instruction and validate it
+ *
+ * @param instruction Transaction instruction to decode
+ * @param programId   SPL Token program account
+ *
+ * @return Decoded, valid instruction
+ */
+export function decodeUnwrapLamportsInstruction(
+    instruction: TransactionInstruction,
+    programId: PublicKey,
+): DecodedUnwrapLamportsInstruction {
+    if (!instruction.programId.equals(programId)) throw new TokenInvalidInstructionProgramError();
+    if (instruction.data.length !== unwrapLamportsInstructionData.span) throw new TokenInvalidInstructionDataError();
```

Comment:
I'm not sure if this works considering the instruction has a variable length, but we can just keep it since it should work in some cases